CONFIGURATION.md

```diff
@@ -0,0 +1,297 @@
+
+`hlx deploy` gets the current git-remote `git remote get-url origin` as `$CURRENT_CODE_REPO` and checks all strains for a `code` property that matches the `$CURRENT_CODE_REPO`.
+
+If no strains are affected, `hlx deploy` will print a new strain config to `stdout` that points to the new `code` location (`$CURRENT_CODE_REPO`), copies all other values from `default` except for `url` or `condition`. The new strain will have an auto-generated, hard-to-guess name, so that it cannot unwittingly be accessed.
+
+When running `hlx deploy --add=foo` the new strain will be added to the configuration file automatically and `hlx deploy` will instead show instructions on accessing the strain.
+
+When running `hlx deploy --add=default` the default strain will be created or updated.
+
+> Notes:
+> - Q: why is it important to have a random name? why not using the branch-name ?
+> - A: Two reasons:
+>      - to avoid conflicts
+>      - to prevent people from forging the X-Strain cookie and getting access to development- or staging-only strains.
+
+A deployment that does not affect any strains will have a non-zero exit code, so that it can fail in CI.
+
+
+
+## Discussion: files
+
+> Notes from the 4/2018 hackathon:
+
+- we will have **one** config helix-config.yaml
+- it will create a canonical `effective-helix-config.yaml` which contains comments for the source of all values
+- the effective file will/must not be checked into git
+
+
+## Discussion: No implicit defaults
+
+I have a hard time reconciling the requirement 
+
+> - the ability to start developing without any configuration file at all
+
+which points at defaults and "magic" values with the kind of predictability and rigidity we need for requirements like
+
+> - the assurance that none of the above will break a published site
+> - the ability to do all of the above in a fully automated fashion on a CI system (e.g. generate and deploy a new strain for a new branch)
+
+### Proposal: No implicit defaults
+
+Therefore, a radical proposal: **we won't have any implicit defaults**. 
+At all. Strains won't magically inherit configuration values from:
+- the global properties in the `helix-config.yaml`
+- the command line
+- application (CLI) provided defaults
+- runtime (Fastly) provided defaults
+
+Unless you configure or reference a configuration value directly, it won't be effective. 
+Unless all required configuration values are set, you won't be able to deploy or publish.
+
+### Proposal: Generate a usable default configuration
+
+The flipside of this is that in order to enable development without any upfront configuration, we need to generate a usable default configuration taking the current context into consideration.
+
+A default configuration should be suggested when:
+- running `hlx deploy`
+- running `hlx publish`
+- running `hlx perf`
+- running `hlx demo`
+- outside of a CI environment
+
+A default configuartion should be saved when:
+- running `hlx * --save`
```

Comment:
As it is the default, I'd say it just overwrites what has been there before (or even refuses to overwrite, so that `--save` only works when there is no `helix-config.yaml`.